compiler-core/src/language_server/annotation_printer.rs

```diff
@@ -0,0 +1,218 @@
+
+    fn name_clashes_if_unqualified(&mut self, type_: &EcoString, module: &str) -> bool {
+        match self.printed_types.get(type_) {
+            None => false,
+            Some(previous_module) if module == previous_module => false,
+            Some(_different_module) => true,
+        }
+    }
+
+    fn args_to_string(&mut self, args: &[Arc<Type>]) -> String {
+        if args.is_empty() {
+            return String::new();
+        }
+
+        let args =
+            Itertools::intersperse(args.iter().map(|t| self.print(t)), ", ".to_string()).collect();
+        args
+    }
+
+    pub fn generic_type_var(&mut self, id: u64) -> String {
+        match self.names.get(&id) {
+            Some(n) => {
+                let _ = self.printed_types.insert(n.clone(), "".into());
+                n.to_string()
+            }
+            None => {
+                let n = self.next_letter();
+                let _ = self.names.insert(id, n.clone());
+                let _ = self.printed_types.insert(n.clone(), "".into());
+                n.to_string()
+            }
+        }
+    }
+
+    fn next_letter(&mut self) -> EcoString {
+        let alphabet_length = 26;
+        let char_offset = 97;
+        let mut chars = vec![];
+        let mut n;
+        let mut rest = self.uid;
+
+        loop {
+            n = rest % alphabet_length;
+            rest /= alphabet_length;
+            chars.push((n as u8 + char_offset) as char);
+
+            if rest == 0 {
+                break;
+            }
+            rest -= 1
+        }
+
+        self.uid += 1;
+        chars.into_iter().rev().collect()
+    }
+}
+
+#[test]
+fn test_type_alias() {
+    let mut type_aliases = HashMap::new();
+    let module_aliases = HashMap::new();
+    let type_parameters = HashMap::new();
+
+    let _ = type_aliases.insert(
+        (EcoString::from("mod"), EcoString::from("Cat")),
+        EcoString::from("c"),
+    );
+
+    let mut printer =
+        AnnotationPrinter::new(&type_aliases, &module_aliases, &type_parameters, "".into());
+
+    let typ = Type::Named {
+        name: EcoString::from("Cat"),
+        args: vec![],
+        module: EcoString::from("mod"),
+        publicity: crate::ast::Publicity::Public,
+        package: EcoString::from(""),
+    };
+
+    assert_eq!(printer.print(&typ), "c");
+}
+
+#[test]
+fn test_type_parameter() {
+    let type_aliases = HashMap::new();
+    let module_aliases = HashMap::new();
+    let mut type_parameters = HashMap::new();
+
+    let type_var = TypeAst::Var(crate::ast::TypeAstVar {
+        name: EcoString::from("T"),
+        location: crate::ast::SrcSpan::default(),
+    });
+
+    let _ = type_parameters.insert(0, type_var);
+
+    let mut printer =
+        AnnotationPrinter::new(&type_aliases, &module_aliases, &type_parameters, "".into());
+
+    let typ = Type::Var {
+        type_: Arc::new(std::cell::RefCell::new(TypeVar::Generic { id: 0 })),
+    };
+
+    assert_eq!(printer.print(&typ), "T");
+}
```

Comment:
Nice! Let's have tests for all the other possible paths through the type printer. We want full coverage to make sure that we are generating correct code.

Reply:
Gotcha. Will add more tests.

Reply:
Are you working on these? 🙏